lib/src/system_cache.dart

```diff
@@ -270,6 +270,26 @@ class SystemCache {
 
     return latest;
   }
+
+  /// Write a README.md file in the root of the cache directory to document the
+  /// contents of the folder.
+  ///
+  void ensureReadme() {
+    final readmePath = p.join(rootDir, 'README.md');
```

Comment:
```suggestion
  }

 var _ensuredReadme = false;

  /// Write a README.md file in the root of the cache directory to document the
  /// contents of the folder.
  ///
  /// This should only be called when we are doing another operation that is
  /// modifying the `PUB_CACHE`. This ensures that users won't experience
  /// permission errors because we writing a `README.md` file, in a flow that
  /// the user expected wouldn't have issues with a read-only `PUB_CACHE`.
  void ensureReadme() {
    // Only ensure README once per execution
    if (_ensuredReadme) {
      return;
    }
    _ensuredReadme = true;
    
    final readmePath = p.join(rootDir, 'README.md');
```